src/meshTools/AMIInterpolation/AMIInterpolation/AMIMethod/directAMI/directAMI.C

```diff
@@ -232,96 +403,107 @@ void Foam::directAMI<SourcePatch, TargetPatch>::calculate
-    DynamicList<label> srcSeeds(10);
+    thrust::transform
+    (
+        srcFaces.begin(),
+        srcFaces.end(),
+        srcCf.begin(),
+        faceCentreFunctor
+        (
+            this->srcPatch_.getFaceNodes().data(),
+            srcPoints.data()
+        )
+    );
 
-    // list to keep track of tgt faces used to seed src faces
-    labelList srcTgtSeed(srcAddr.size(), -1);
-    srcTgtSeed[srcFaceI] = tgtFaceI;
+    labelgpuList srcCount(this->srcPatch_.size(), 0);
+    labelgpuList tgtCount(this->tgtPatch_.size(), 0);
 
-    // list to keep track of whether src face can be mapped
-    // 1 = mapped, 0 = untested, -1 = cannot map
-    labelList mapFlag(srcAddr.size(), 0);
+    gpuList<labelList> srcAddr(this->srcPatch_.size(), labelList(15, -1));
+    gpuList<labelList> tgtAddr(this->tgtPatch_.size(), labelList(15, -1));
 
-    label nTested = 0;
-    DynamicList<label> nonOverlapFaces;
-    do
+    for (int i = 0; i < srcFaces.size(); ++i)
     {
-        srcAddr[srcFaceI].append(tgtFaceI);
-        tgtAddr[tgtFaceI].append(srcFaceI);
-
-        mapFlag[srcFaceI] = 1;
-
-        nTested++;
-
-        // Do advancing front starting from srcFaceI, tgtFaceI
-        appendToDirectSeeds
+        thrust::counting_iterator<unsigned> tgtStart(0);
+        thrust::counting_iterator<unsigned> tgtEnd = tgtStart + tgtFaces.size();
+        thrust::constant_iterator<unsigned> srcStart(i);
+        thrust::transform
         (
-            mapFlag,
-            srcTgtSeed,
-            srcSeeds,
-            nonOverlapFaces,
-            srcFaceI,
-            tgtFaceI
+            tgtStart,
+            tgtEnd,
+            srcStart,
+            ShootRayFunctor
+            (
+                tgtPoints.data(),
+                srcPoints.data(),
+                srcCf.data(),
+                tgtFaces.data(),
+                srcFaces.data(),
+                srcCount.data(),
+                tgtCount.data(),
+                srcAddr.data(),
+                tgtAddr.data()
+            )
         );
-
-        if (srcFaceI < 0 && nTested < this->srcPatch_.size())
-        {
-            restartAdvancingFront(mapFlag, nonOverlapFaces, srcFaceI, tgtFaceI);
-        }
-
-    } while (srcFaceI >= 0);
-
-    if (nonOverlapFaces.size() != 0)
-    {
-        Pout<< "    AMI: " << nonOverlapFaces.size()
-            << " non-overlap faces identified"
-            << endl;
-
-        this->srcNonOverlap_.transfer(nonOverlapFaces);
     }
 
-    // transfer data to persistent storage
-    forAll(srcAddr, i)
+    for (int i = 0; i < this->srcPatch_.size(); ++i)
     {
         scalar magSf = this->srcMagSf_[i];
-//        srcWeights[i] = scalarList(srcAddr[i].size(), magSf);
         srcWeights[i] = scalarList(1, magSf);
-        srcAddress[i].transfer(srcAddr[i]);
+
+        labelList l = srcAddr.get(i);
+        l.setSize(srcCount.get(i));
+        srcAddress[i].transfer(l);
```

Comment:
Otherwise, how do we update the `srcAddress` and `tgtAddress` arrays? I would love to avoid doing this as well.